File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.go
func subarraySum(nums []int, k int) int {
    /*
    [1,2,3,4,5,5,7] k=10
    prefix[1,3,6,10,15,20,28]
    if prefix-k in hashmap, means there is an subarray sum==k
    if not, there is not subarray
    when we need subarray, prefix is a very good choice
    */
    res:=map[int]int{0:1}
    prefix:=0
    ans:=0
    for _, i := range nums{
        prefix+=i
        ans+=res[prefix-k]
        res[prefix]++
    }
    return ans 
}

File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.py
class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        res=defaultdict(int)
        prefix=0
        ans=0
        for i in range(len(nums)):
            # 一定要有res[0],否则prefix=k无法计入
            res[prefix]+=1
            prefix+=nums[i]
            ans+=res[prefix-k]
        return ans
        